### PR TITLE
fix #4610: Add input arg descriptions to sirepo.api.Spec

### LINKED ISSUE #4610: Add input arg descriptions to sirepo.api.Spec
See radiasoft/pykern/issues/205

### PR DESCRIPTION
**Pull request by rorour**

### CODE REVIEW

File: sirepo/job_api.py [hunk 1]
```diff
@@ -49,7 +49,13 @@ def api_admJobs(self):
     def api_analysisJob(self):
         return self.request()
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user",
+        sid="SimId",
+        model="Model",
+        frame="Frame",
+        suffix="String optional",
```
Comment (robnagler) [inline sirepo/job_api.py:57]: I think names like Model and Frame are too generic. Model l means two things: the actual data and the model name, which is confusing. In this case, I think this is a ComputeModel, but I could be wrong. @e-carlin what say you?

Frame is a number. It really is an index, so maybe DataFileIndex.

String should be constrained. At is a FileSuffix, and it should be limited to 30 characters or something like that.

File: sirepo/job_api.py [hunk 2]
```diff
@@ -135,7 +141,7 @@ def api_runCancel(self):
         # Always true from the client's perspective
         return self.reply_json({"state": "canceled"})
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="PKDict all_input")
```
Comment (robnagler) [inline sirepo/job_api.py:144]: `data="SimData all_input"`.

File: sirepo/job_api.py [hunk 3]
```diff
@@ -174,7 +180,7 @@ def api_sbatchLogin(self):
         r.sbatchCredentials = r.pkdel("data")
         return self.request(_request_content=r)
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", frame_id="FrameId")
```
Comment (robnagler) [inline sirepo/job_api.py:183]: SimFrameId.

File: sirepo/server.py [hunk 4]
```diff
@@ -104,7 +104,7 @@ def api_copySimulation(self):
         )
         return self._save_new_and_reply(req, d)
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", filename="FileName", file_type="FileType")
```
Comment (robnagler) [inline sirepo/server.py:107]: SimFileName and SimFileType

File: sirepo/server.py [hunk 5]
```diff
@@ -526,7 +535,7 @@ def api_srUnit(self):
         v.op()
         return self.reply_ok()
 
-    @sirepo.api.Spec("allow_visitor")
+    @sirepo.api.Spec("allow_visitor", path_info="PathInfo")
```
Comment (robnagler) [inline sirepo/server.py:538]: FilePath

File: sirepo/server.py [hunk 6]
```diff
@@ -584,7 +593,14 @@ def api_updateFolder(self):
             simulation_db.save_simulation_json(r, fixup=False)
         return self.reply_ok()
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user",
+        sid="SimId",
+        file_type="FileType",
```
Comment (robnagler) [inline sirepo/server.py:599]: Let's keep the arguments in order. as they are defined.

File: sirepo/sim_api/jupyterhublogin.py [hunk 7]
```diff
@@ -46,7 +46,7 @@ def api_checkAuthJupyterhub(self):
             u = create_user()
         return self.reply_ok(PKDict(username=u))
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="PKDict all_input")
```
Comment (robnagler) [inline sirepo/sim_api/jupyterhublogin.py:49]: SimData

File: sirepo/srtime.py [hunk 8]
```diff
@@ -47,14 +47,13 @@ def adjust_time(days, sapi=None):
 
 
 class API(sirepo.api.Base):
-    @sirepo.api.Spec("internal_test")
+    @sirepo.api.Spec("internal_test", days="String optional")
```
Comment (robnagler) [inline sirepo/srtime.py:50]: Read the docstring and then specify a type.

File: sirepo/api.py [hunk 9]
```diff
@@ -63,8 +63,9 @@ def reply_static_jinja(self, base, ext, j2_ctx, cache_ok=False):
 
 
 class Spec:
-    def __init__(self, perm):
+    def __init__(self, perm, **kwargs):
         self.perm = perm
+        self.args = kwargs
```
Comment (robnagler) [inline sirepo/api.py:68]: keep names consistent and convert to PKDict

File: sirepo/auth/__init__.py [hunk 10]
```diff
@@ -85,7 +85,7 @@
 
 
 class API(sirepo.api.Base):
-    @sirepo.api.Spec("require_cookie_sentinel")
+    @sirepo.api.Spec("require_cookie_sentinel", display_name="DisplayName")
```
Comment (robnagler) [inline sirepo/auth/__init__.py:88]: UserDisplayName

File: sirepo/auth/email.py [hunk 11]
```diff
@@ -51,7 +51,7 @@
 
 
 class API(sirepo.api.Base):
-    @sirepo.api.Spec("allow_cookieless_set_user")
+    @sirepo.api.Spec("allow_cookieless_set_user", token="Token")
```
Comment (robnagler) [inline sirepo/auth/email.py:54]: EmailAuthToken

File: sirepo/auth_role_moderation.py [hunk 12]
```diff
@@ -34,7 +34,7 @@
 
 
 class API(sirepo.api.Base):
-    @sirepo.api.Spec("require_adm")
+    @sirepo.api.Spec("require_adm", token="Token", status="ModerationStatus")
```
Comment (robnagler) [inline sirepo/auth_role_moderation.py:37]: AuthModerationStatus

File: sirepo/auth_role_moderation.py [hunk 13]
```diff
@@ -102,7 +102,7 @@ def api_getModerationRequestRows(self):
             ),
         )
 
-    @sirepo.api.Spec("allow_sim_typeless_require_email_user")
+    @sirepo.api.Spec("allow_sim_typeless_require_email_user", reason="String")
```
Comment (robnagler) [inline sirepo/auth_role_moderation.py:105]: AuthModerationReason

File: sirepo/comsol_register.py [hunk 14]
```diff
@@ -23,10 +23,8 @@ def api_comsol(self):
             "https://www.radiasoft.net/services/comsol-certified-consulting/",
         )
 
-    @sirepo.api.Spec("allow_visitor")
+    @sirepo.api.Spec("allow_visitor", name="Name", email="Email")
```
Comment (robnagler) [inline sirepo/comsol_register.py:26]: UserDisplayName

File: sirepo/job_api.py [hunk 15]
```diff
@@ -31,15 +31,15 @@
 
 
 class API(sirepo.api.Base):
-    @sirepo.api.Spec("internal_test")
+    @sirepo.api.Spec("internal_test", days="Days")
```
Comment (robnagler) [inline sirepo/job_api.py:34]: TimeDeltaDays

**Review by robnagler — CHANGES_REQUESTED** [review]
All types should be constrainable. No unlimited types like String. PKDict is not a "type" so it should SimData or maybe a new type, depending on the context. I caught many but others may need fixing up.

File: sirepo/job_api.py (hunk 1 above)
Comment (e-carlin) [reply sirepo/job_api.py:57]: Wouldn't that be an AnalysisModel? The user may be clicking the downloadDataFile on the fooAnimation panel  and underneath the computeModel may be something else. The model that comes in I thought was the AnalysisModel.

Comment (rorour) [reply sirepo/job_api.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/job_api.py (hunk 2 above)
Comment (rorour) [reply sirepo/job_api.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/job_api.py (hunk 3 above)
Comment (rorour) [reply sirepo/job_api.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/server.py (hunk 4 above)
Comment (rorour) [reply sirepo/server.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/server.py (hunk 5 above)
Comment (rorour) [reply sirepo/server.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/server.py (hunk 6 above)
Comment (rorour) [reply sirepo/server.py:599]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/auth/__init__.py (hunk 10 above)
Comment (rorour) [reply sirepo/auth/__init__.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/auth_role_moderation.py (hunk 12 above)
Comment (rorour) [reply sirepo/auth_role_moderation.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/auth_role_moderation.py (hunk 13 above)
Comment (rorour) [reply sirepo/auth_role_moderation.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/comsol_register.py (hunk 14 above)
Comment (rorour) [reply sirepo/comsol_register.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/job_api.py (hunk 15 above)
Comment (rorour) [reply sirepo/job_api.py]: 24fd189959a5d6096a9482e74f1ec55c9faeb6de

File: sirepo/server.py [hunk 16]
```diff
@@ -251,7 +256,7 @@ def api_findByNameWithAuth(
             query=m.includeMode and PKDict(application_mode=application_mode),
         )
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", filename="SimFileName", data="SimData all_input")
```
Comment (robnagler) [inline sirepo/server.py:259]: data is not SimData. This is a weird method. call it ApplicationDataSpec (no need for all_input) or filename. That will be available from ApplicationDataSpec. Change from data to spec.

File: sirepo/server.py [hunk 17]
```diff
@@ -276,7 +281,7 @@ def api_getApplicationData(self, filename=None):
                 )
             return self.reply_json(res)
 
-    @sirepo.api.Spec("allow_cookieless_require_user")
+    @sirepo.api.Spec("allow_cookieless_require_user", sim_data="SimData")
```
Comment (robnagler) [inline sirepo/server.py:284]: zip=ImportArchiveZip

File: sirepo/server.py [hunk 18]
```diff
@@ -372,13 +377,15 @@ def s(data):
...
-    @sirepo.api.Spec("allow_visitor")
+    @sirepo.api.Spec("allow_visitor", path_info="PathInfo optional")
     def api_homePage(self, path_info=None):
         return self.call_api(
             "staticFile", kwargs=PKDict(path_info="en/" + (path_info or "landing.html"))
         )
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user", sid="SimId", model="Model optional", title="NbTitle"
```
Comment (robnagler) [inline sirepo/server.py:387]: DownloadNamePostfix optional

File: sirepo/server.py [hunk 19]
```diff
@@ -406,7 +413,7 @@ def api_newSimulation(self):
             req.template.new_simulation(d, req.req_data)
         return self._save_new_and_reply(req, d)
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", sid="SimId", model="Model", title="Title")
```
Comment (robnagler) [inline sirepo/server.py:416]: model=ComputeModelName optional, title=DownloadNamePostfix optional, sid has to be simulation_id to match args.

File: sirepo/server.py [hunk 20]
```diff
@@ -482,7 +491,7 @@ def api_simulationData(
                 e.sr_response["redirect"]["section"] = section
             return http_reply.headers_for_no_cache(self.reply_json(e.sr_response))
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="SimData all_input")
```
Comment (robnagler) [inline sirepo/server.py:494]: search=SearchSpec

File: sirepo/server.py [hunk 21]
```diff
@@ -552,7 +561,7 @@ def api_staticFile(self, path_info=None):
             return self.reply_html(p)
         return flask.send_file(p, conditional=True)
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="SimData all_input")
     def api_updateFolder(self):
         # TODO(robnagler) Folder should have a serial, or should it be on data
         req = self.parse_post()
```
Comment (robnagler) [inline sirepo/server.py:567]: oldName=SimFolderPath newName=SimFolderPath

File: sirepo/sim_api/jupyterhublogin.py [hunk 22]
```diff
@@ -46,7 +46,7 @@ def api_checkAuthJupyterhub(self):
             u = create_user()
         return self.reply_ok(PKDict(username=u))
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="SimData all_input")
```
Comment (robnagler) [inline sirepo/sim_api/jupyterhublogin.py:49]: do_migration=Bool sim_type=f"SimType default={_SIM_TYPE}"

File: sirepo/job_api.py [hunk 23]
```diff
@@ -135,7 +141,7 @@ def api_runCancel(self):
         # Always true from the client's perspective
         return self.reply_json({"state": "canceled"})
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="SimData all_input")
```
Comment (robnagler) [inline sirepo/job_api.py:144]: RunMultiSpec (no need for all_input)

File: sirepo/server.py [hunk 24]
```diff
@@ -140,7 +142,9 @@ def api_downloadFile(self, simulation_type, simulation_id, filename):
                 sirepo.util.raise_not_found("lib_file={} not found", p)
             raise
 
-    @sirepo.api.Spec("allow_visitor")
+    @sirepo.api.Spec(
+        "allow_visitor", error="String all_input", client_ip="IPAddress optional"
```
Comment (robnagler) [inline sirepo/server.py:146]: spec=ErrorLoggingSpec (no need for all_input). I think we don't need client_ip here. What we'll do is ErrorLoggingSpec will know that it has to extract client_ip from the request.

>

File: sirepo/server.py [hunk 25]
```diff
@@ -91,7 +91,7 @@ def api_copyNonSessionSimulation(self):
         #     req.template.copy_related_files(data, str(src), str(target))
         return res
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", sid="SimId", folder="FolderName", name="SimName")
```
Comment (robnagler) [inline sirepo/server.py:94]: SimFolderName

File: sirepo/server.py [hunk 26]
```diff
@@ -293,7 +298,7 @@ def api_importArchive(self):
             PKDict(simulationId=data.models.simulation.simulationId),
         )
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", file="File", sid="SimId")
     def api_importFile(self, simulation_type):
         """
         Args:
```
Comment (robnagler) [inline sirepo/server.py:304]: file=ImportFile folder=SimFolderPath sid=SimId arguments=ImportArgs optional

File: sirepo/server.py [hunk 27]
```diff
@@ -584,7 +593,14 @@ def api_updateFolder(self):
             simulation_db.save_simulation_json(r, fixup=False)
         return self.reply_ok()
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user",
+        file="FileArg input",
+        file_type="FileType",
```
Comment (robnagler) [inline sirepo/server.py:599]: LibFileType

File: sirepo/server.py [hunk 28]
```diff
@@ -584,7 +593,14 @@ def api_updateFolder(self):
             simulation_db.save_simulation_json(r, fixup=False)
         return self.reply_ok()
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user",
+        file="FileArg input",
```
Comment (robnagler) [inline sirepo/server.py:598]: LibFile (no need for input)

File: sirepo/server.py [hunk 29]
```diff
@@ -584,7 +593,14 @@ def api_updateFolder(self):
...
+        file_type="FileType",
+        filename="SimFileName",
```
Comment (robnagler) [inline sirepo/server.py:600]: No need. This comes directly from the files["file"] value

File: sirepo/server.py [hunk 30]
```diff
@@ -584,7 +593,14 @@ def api_updateFolder(self):
...
+        sid="SimId",
+        sim_data="SimData",
```
Comment (robnagler) [inline sirepo/server.py:602]: There is no sim_data.

Add confirm=Bool optional

**Review by robnagler — CHANGES_REQUESTED** [review]
Try to understand the difference between inputs and values that are computed from inputs.

File: sirepo/job_api.py [hunk 31]
```diff
@@ -118,7 +124,7 @@ def api_jobSupervisorPing(self):
             e = "unexpected exception"
         return PKDict(state="error", error=e)
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", data="PKDict all_input")
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/job_api.py:127]: Will the contents of **self.parse_post() come from self.kwargs now?

File: sirepo/server.py (hunk 28 above)
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 27 above)
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 26 above)
Comment (rorour) [reply sirepo/server.py:317]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 25 above)
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/job_api.py (hunk 23 above)
Comment (rorour) [reply sirepo/job_api.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/sim_api/jupyterhublogin.py (hunk 22 above)
Comment (rorour) [reply sirepo/sim_api/jupyterhublogin.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 20 above)
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 21 above)
Comment (rorour) [reply sirepo/server.py:588]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py (hunk 16 above)
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py [hunk 32]
```diff
@@ -183,15 +187,19 @@ def api_favicon(self):
     def api_forbidden(self):
         sirepo.util.raise_forbidden("app forced forbidden")
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec(
+        "require_user", sid="SimId optional", file_type="FileType", sim_data="SimData"
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/server.py [hunk 33]
```diff
@@ -158,7 +162,7 @@ def api_errorLogging(self):
             )
         return self.reply_ok()
 
-    @sirepo.api.Spec("require_user")
+    @sirepo.api.Spec("require_user", sid="SimId", filename="SimFileName")
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/server.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/job_api.py [hunk 34]
```diff
@@ -31,15 +31,15 @@
...
+    @sirepo.api.Spec("internal_test", days="TimeDeltaDays")
     def api_adjustSupervisorSrtime(self, days):
         return self.request(
             api_name="not used",
             _request_content=PKDict(days=days),
             _request_uri=self._supervisor_uri(sirepo.job.SERVER_SRTIME_URI),
         )
 
-    @sirepo.api.Spec("require_adm")
+    @sirepo.api.Spec("require_adm", data="PKDict all_input")
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/job_api.py]: 9d450384bbf20294b6007b678c3dde0143022f40

File: sirepo/job_api.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/job_api.py]: 9d450384bbf20294b6007b678c3dde0143022f40

**Comment by rorour** [comment]
@robnagler have you gotten a chance to look at this?

**Review by robnagler — APPROVED** [review]
Looks good

**Pull request merged**